### DISCUSSION TITLE
Support for GD32C103xx?

### DISCUSSION THREAD

**Discussion by RudolphRiedel** [CommunityGD32Cores/platform-gd32#30] [General]
Hi,

I just designed and ordered a bunch of boards with GD32C103CB.
This is for an open source project and I will release the design files if everything is up and running,
for now a couple of things are a bit uncertain.

Anyways, I won't be doing the firmware but I would like to play with the board as well.
And since I prefer PlatformIO I checked support for GD32 in PIO first and ended up here.

The examples are compiling just fine, I only had to manually install GD32 as PIO could not find it automatically.

Next I adapted one of the board files: 
[genericGD32C103CB.zip](https://github.com/CommunityGD32Cores/platform-gd32/files/8838500/genericGD32C103CB.zip)

Then I added the board file to the platformio.ini of gd32-spl-blinky and copied over GD32C10x.svd.
Now it starts to compile but fails and I have to admit that I need help.

Please add support for GD32C103.

**Comment by maxgerhardt** [comment]
Interesting, I didn't even know the GD32C10x series **existed**. There's no news mention of it in https://www.gigadevice.com/about/press-room/. Only going to Cortex-M4 page now there are these microcontrollers which I've never heard before previously.

The correct integration is downloading the their firmware libary from https://gd32mcu.com/en/download/0?kw=GD32C1 and adapting https://github.com/CommunityGD32Cores/gd32-pio-spl-package accordingly (with possible build script mods here). We will look into adding support for all GD32C10x chips through automated board file generates, same as we did with all other chips.

**Reply by maxgerhardt** [reply]
Tracked in https://github.com/CommunityGD32Cores/platform-gd32/issues/31.

**Reply by RudolphRiedel** [reply]
Nice, that sounds very promising. :-)
I did learn about the chip by pure coincidence earlier this week and since JLCPCB has some in stock I decided to give it a try.
There is not much in stock currently of anything, even less so when i comes to controllers with USB and two CAN-FD interfaces.

**Reply by maxgerhardt** [reply]
I've added support for C10x in https://github.com/CommunityGD32Cores/platform-gd32/commit/bedc0c0247d5a8d6996fe6aa5198c940392ce701 and https://github.com/CommunityGD32Cores/gd32-pio-spl-package/commit/33f193fcd1a713b7f4cc98fac0a0a86251ebad4a.

Could you please delete `C:\Users\<user>\.platformio\platforms\gd32` again and reinstall it per README? Then use the updated [gd32-spl-blinky](https://github.com/CommunityGD32Cores/platform-gd32/tree/main/examples/gd32-spl-blinky) example with the `genericGD32C103CB` environment selected in the [project environment switcher](https://docs.platformio.org/en/latest/integration/ide/vscode.html#project-tasks)? Does it build? Is it able to upload? I'm not sure about the latter, I'm using STM32F4 openocd files here since that's the closest to a Cortex-M4.

**Reply by maxgerhardt** [reply]
My bad, depending on whether your board has a 8MHz crystal or not, the defaults that GigaDevice set in their SPL code will not work. Let me not make assumptions on whether and what highspeed crystal you have on your board. Please add
```ini
build_flags = -D__SYSTEM_CLOCK_120M_PLL_IRC8M=120000000U
```
at the bottom of the `platformio.ini` to switch the clock to internal 8MHz RC + PLL to 120MHz. Otherwise it tries to PLL a 8MHz external crystal.

**Reply by RudolphRiedel** [reply]
Wow, that was fast, thank you!
I had to delete framework-spl-gd32 from /.platformio/packages as well but after that the example compiles just fine. :-)

My board will be using an external 16MHz crystal, or at least this is the plan.
So I guess I just need to change the HXTAL_VALUE in the genericGD32C103CB.json to 16000000?

Unfortunately the boards are at least two weeks out...

**Reply by maxgerhardt** [reply]
>My board will be using an external 16MHz crystal, or at least this is the plan.

Mmmh, sadly Gigadevice's clock code is not flexible and doesn't recognize any other values than 8MHz or 25MHz ([see here](https://github.com/CommunityGD32Cores/gd32-pio-spl-package/blob/main/gd32/cmsis/variants/gd32c10x/system_gd32c10x.c#L760-L766)), 16MHz would need a slight adapation (aka duplicate the 8MHz branch but half one multiplier). For the very first test I'd recommend using the internal clock per instructions above, we can expand on Gigadevice's shortcomings in the clock code later and add more support. 

In the meantime I'll be working on more Arduino-GD32 support.

**Reply by RudolphRiedel** [reply]
Thank you for the heads-up!
That is not much of an issue, I need to populate the crystals by hand anyways and I do have 8MHz crystals with the same footprint.

**Reply by maxgerhardt** [reply]
I'm curious, were there any updates? :)

**Reply by RudolphRiedel** [reply]
Any? Well, yes. :-)
I got one board to work and even expanded the scope of my initial hardware test to use SPI over DMA.
So things look really good at least in regards of PlatformIO and the GD32 core.
Even more so after Segger kindly added the GD32C to the supported controllers when I asked them.

It's not looking as good for our application though.
The first issue was that I was getting enumeration errors under windows with the builtin bootloader.
I had to switch to a 25MHz crystal to fix that, apparently the bootloader is trying 25MHz first.
Then we found out that the GDC32C103 USB is not providing enough endpoints for our application.

Well, that is that, things are kind of on hold right now but this is completely unrelated to GD32 core. :-)

**Comment by Isaacloss** [comment]
Olá, o chip da minha placa é o LCM32F037 k6T8. Consigo alterar o firmware para ele?

Obrigado!
